proposals/p2173.md

```diff
@@ -313,49 +317,165 @@ proposal suggests a simpler rule:
 
 It's possible for one `=` constraint in a `where` to refer to another. When this
 happens, the constraint `C where A and B` is interpreted as
-`(C where A) where B`, so rewrites in `A` are applied to names in `B`, but
-rewrites in `B` are not applied to names in `A`:
+`(C where A) where B`, so rewrites in `A` are applied immediately to names in
+`B`, but rewrites in `B` are not applied to names in `A` until the constraint is
+[resolved](#constraint-resolution):
 
 ```
 interface C {
   let T:! Type;
   let U:! Type;
+  let V:! Type;
 }
-// ✅ Same as `(C where .T = i32) where .U = .T`, which is
-// the same as `(C where .T = i32) where .U = i32`, which is
-// the same as `C where .T = i32 and .U = i32`.
-fn F[A:! C where .T = i32 and .U = .T]() {}
+class M {
+  alias Me = Self;
+}
+// ✅ Same as `C where .T = M and .U = M.Me`, which is
+// the same as `C where .T = M and .U = M`.
+fn F[A:! C where .T = M and .U = .T.Me]() {}
+// ❌ No member `Me` in `A.T:! Type`.
+fn F[A:! C where .U = .T.Me and .T = M]() {}
 ```
 
-If a constraint is formed where a rewrite would apply to the right-hand side of
-another rewrite in the same constraint, but is not applied because it was not in
-scope, the program is invalid.
+### Combining constraints with `extends`
+
+Within an interface or named constraint, `extends` can be used to extend a
+constraint that has rewrites.
 
 ```
-// ❌ Rewritten member `.U` is mentioned in a context
-// where its rewrite is not in scope.
-fn BadG[A:! C where .T = .U and .U = i32]() {}
-// ❌ Rewritten member `.U` is mentioned in a context
-// where its rewrite is not in scope.
-fn BadH[A:! C where .T = .U and .U = .T]() {}
-// ❌ Rewritten member `.T` is mentioned in a context
-// where its rewrite is not in scope.
-fn BadI[A:! (C where .T = i32) & (C where .U = .T)]() {}
+interface A {
+  let T:! Type;
+  let U:! Type;
+}
+interface B {
+  extends A where .T = .U and .U = i32;
+}
+
+var n: i32;
+
+// ✅ Resolved constraint on `T` is
+// `B where .(A.T) = i32 and .(A.U) = i32`.
+// `T.(A.T)` is rewritten to `i32`.
+fn F(T:! B) -> T.(A.T) { return n; }
 ```
 
-In more detail, what happens here is:
+### Combining constraints with `impl as`
 
-1.  The constraint is fully elaborated. `.X` is replaced with `.Self.X`. For
-    example, `C where .T = .U and .U = i32` is rewritten to
-    `C where .T = .Self.U and .U = i32`.
-2.  When the constraint is applied to a type `V`, the substitution `.Self = V`
-    is performed. The type of type used for `V` is
-    `C where .T = <error> and .U = <error>`.
-3.  During this substitution, if either `<error>` rewrite rule would be used,
-    the result is an error. Substitution into `.Self.U` produces `V.U`, giving
-    an error.
+Within an interface or named constraint, the `impl T as C` syntax permits `=`
+constraints. However, because these constraints don't change the type of `T`,
+rewrite constraints will never result in a rewrite being performed, and instead
+are equivalent to `==` constraints.
```

Comment:
Could you go into more detail about what goes wrong if they were still treated as rewrites? Is this a case where we don't know to look for the rewrites, we don't know the order in which rewrites should be applied, or something else?

I'm worried since this is an obstacle to eventually simplifying by dropping support for `==` constraints entirely if we can make `=` constraints cover all needed use cases.

Reply:
The problem is that they're not part of the type of the thing they're rewriting, and only looking in the type of the thing on the left of the `.` for rewrites is the basis for the model and the foundation of the termination argument.

We *could* permit this for `impl as` constraints, but not for `impl SomeOtherType as` constraints, and it seemed more consistent to me to draw the line at the difference between internal (affecting-the-interface-of-`Self`) implementations and external (not-affecting-the-interface-of-`Self`) implementations.

Perhaps we should talk through this more on Discord / open discussion. Maybe there's a way to make this work.

Reply:
Could you add the explanation about `impl SomeOtherType as` to the proposal? I can accept the limitation in the `SomeOtherType != Self` case. I think there is a legitimate question about the `impl as` form. I'm fine with either preliminary decision about what side of the line that form ends up on, but I think we should acknowledge that there are two alternatives under consideration and that we might revisit that specific decision in the future with more experience.

Reply:
I've added a section to "Alternatives considered" on this.